internal/rotate_config.go

```diff
@@ -26,6 +28,9 @@ func (s fullCfgStep) Execute(handler *certRotationContext) error {
 	// Open/decrypt full config with current key
 	config, err := UnmarshallFile(handler.crypto, handler.app.EncryptedConfig, true)
 	if err != nil {
+		if errors.Is(err, os.ErrNotExist) { // os.IsNotExist does not work on wrapped errors
```

Comment:
shouldn't we at least have an empty file here so that we know its checked in with the server at least once? My fear is this scenario:

 * Device configuration exists on server
 * There is no config.encrypted file
 * We do a rotation
 * Device configuration is not encrypted with the wrong key?

Reply:
When there is no config - fioconfig does not create the `config.encrypted` file.
This is exactly why I've got this bug during my tests.

Also, an empty file is not an option, as it is not a valid JSON... we'd need to put _something_ into it.

That said, I don't want a certificate rotation to alter the filesystem in an unexpected way.
I mean, if there was no file - we should not create the one.

Note: encrypting the server-side device config is a separate step, which has nothing to do with `config.encrypted`.

Reply:
To put it on more practical way:

- If the cert rotation was triggered from the config file, the config.encrypted is there, so it does not come this code path.
- This fix is for a use case when the cert rotation was triggered by a user explicitly running the `fioconfig renew-client-cert` command.
  Note: the user might even want to use a separate tool for their config management, and only call fioconfig for cert rotation.

Reply:
by "empty file" I mean with empty json content `{}`.  I'm not implying we should create one if it doesn't exist. I want to make sure that in the event the file isn't there, but there's configuration on the server (ie. it somehow got deleted on the device) - that we aren't going to break something. Sounds like this isn't the case as cert rotation checks from the server before proceeding.

Do we write out an empty json file when we get the 204 from the device-gateway now or no file?

Reply:
> To put it on more practical way:
> 
>     * If the cert rotation was triggered from the config file, the config.encrypted is there, so it does not come this code path.
> 
>     * This fix is for a use case when the cert rotation was triggered by a user explicitly running the `fioconfig renew-client-cert` command.
>       Note: the user might even want to use a separate tool for their config management, and only call fioconfig for cert rotation.

Maybe we could skip the overall rotate config step if there is no config?

Reply:
> Maybe we could skip the overall rotate config step if there is no config?

This is actually what this fix does.

Reply:
I don't think so, the step is executed, the fix just makes the step do nothing :).

Reply:
OK, so if you mean to remove this step altogether, than it is not what I'd like to make.

The cert rotation can take time and many retries.
Having a step which _checks for config presence_ is more reliable in this unreliable environment than _having no step at all_.
IMHO _checking for config presence_ is not equivalent to _doing nothing_.

Reply:
Not exactly, I thought about adding or skipping this step dynamically depending on the presence of a config.

Anyway, I just proposed an idea, most likely it is not feasible in the given design.

Reply:
I _think_ I understand what you mean by _dynamicity_ here.
Yeah, I think it would require a moderate low-level re-design to get exactly the same high-level behavior.
... Which is synonymous with _not feasible_ in this context.

Reply:
> Do we write out an empty json file when we get the 204 from the device-gateway now or no file?

@doanac during my tests I've seen that we don't write config to disk if the server returns no config.

This is cross-proven by the code inspection:
- Status code 200 - write a file: https://github.com/foundriesio/fioconfig/blob/main/internal/app.go#L349
- Status code 204 - write no file: https://github.com/foundriesio/fioconfig/blob/main/internal/app.go#L366

If you think we can do better, I'd rather discuss this in a separate PR.